src/model.cpp

```diff
@@ -0,0 +1,250 @@
+    current_events().resize(num_groups());
+    future_events().resize(num_groups());
+}
+
+void model::reset() {
+    t_ = 0.;
+    for (auto& group: cell_groups_) {
+        group->reset();
+    }
+
+    communicator_.reset();
+
+    for(auto& q : current_events()) {
+        q.clear();
+    }
+    for(auto& q : future_events()) {
+        q.clear();
+    }
+
+    current_spikes().clear();
+    previous_spikes().clear();
+
+    util::profilers_restart();
+}
+
+time_type model::run(time_type tfinal, time_type dt) {
+    // Calculate the size of the largest possible time integration interval
+    // before communication of spikes is required.
+    // If spike exchange and cell update are serialized, this is the
+    // minimum delay of the network, however we use half this period
+    // to overlap communication and computation.
+    time_type t_interval = communicator_.min_delay()/2;
+
+    time_type tuntil;
+
+    // task that updates cell state in parallel.
+    auto update_cells = [&] () {
+        threading::parallel_for::apply(
+            0u, cell_groups_.size(),
+             [&](unsigned i) {
+                auto &group = cell_groups_[i];
+
+                PE("stepping","events");
+                group->enqueue_events(current_events()[i]);
+                PL();
+
+                group->advance(tuntil, dt);
+
+                PE("events");
+                current_spikes().insert(group->spikes());
+                group->clear_spikes();
+                PL(2);
+            });
+    };
+
+    // task that performs spike exchange with the spikes generated in
+    // the previous integration period, generating the postsynaptic
+    // events that must be delivered at the start of the next
+    // integration period at the latest.
+    auto exchange = [&] () {
+        PE("stepping", "communication");
+
+        PE("exchange");
+        auto local_spikes = previous_spikes().gather();
+        auto global_spikes = communicator_.exchange(local_spikes);
+        PL();
+
+        PE("spike output");
+        local_export_callback_(local_spikes);
+        global_export_callback_(global_spikes.values());
+        PL();
+
+        PE("events");
+        future_events() = communicator_.make_event_queues(global_spikes);
+        PL();
+
+        PL(2);
+    };
+
+    while (t_<tfinal) {
+        tuntil = std::min(t_+t_interval, tfinal);
+
+        event_queues_.exchange();
+        local_spikes_.exchange();
+
+        // empty the spike buffers for the current integration period.
+        // these buffers will store the new spikes generated in update_cells.
+        current_spikes().clear();
+
+        // run the tasks, overlapping if the threading model and number of
+        // available threads permits it.
+        threading::task_group g;
+        g.run(exchange);
+        g.run(update_cells);
+        g.wait();
+
+        t_ = tuntil;
+    }
+
+    // Run the exchange one last time to ensure that all spikes are output
+    // to file.
```

Comment:
Spikes are output to file after each exchange.
We perform the exchange every `min_delay`ms.
When time stepping finishes half way through one of these `min_delay` periods, you have to ensure that the spikes that have been stored in local buffers are also output to file, which requires one last spike exchange.
This was a long standing bug, whereby some spikes at the end of a simulation were not being output.